### PR TITLE
Fix class constructor regression

### LINKED ISSUE #13102: [vim9script] Regression after fixing constructor argument type checking bug
### Steps to reproduce

Patch 9.0.1724 (“vim9class constructor argument type checking bug”) has introduced a regression in one of my Vim 9 libraries. I have [a test suite that caught that](https://github.com/lifepillar/vim-devel/blob/main/start/librelalg/test/test_librelalg.vim), but I was able to nail down the issue to the following self-contained script:

```vim
vim9script

def ListifyKeys(keys: any): list<list<string>>
  if type(keys) == v:t_string
    return [[keys]]
  endif
  return keys
enddef

class Rel
  this.keys: list<list<string>>

  def new(keys: any)
    const keys_: list<list<string>> = ListifyKeys(keys)
  enddef
endclass

def Test()
  var R = Rel.new('A')
enddef

Test()
```

This fails with `E1013: Argument 1: type mismatch, expected list<list<string>> but got string`.

Interestingly, if the `this.keys` attribute is removed, no error occurs.

**Edit:** no error occurs if `R` is created outside the function, either.

### Expected behaviour

I would expect the script above to run without errors. That was the case before patch 9.0.1724.

### Version of Vim

9.0.1899

### Environment

macOS 13.5.1
Apple Terminal
xterm-256color
zsh 5.9

### Logs and stack traces

_No response_

### PR DESCRIPTION
**Pull request by h-east**
Close #13102

### CODE REVIEW

**Comment by h-east** [comment]
change point
- Stopped referencing the function body in compile time
  (When check the function body, there is no end to it.)
- Added TTFLAG_ARG_OBJM for tt_flags. Set when the constructor's formal argument is an object member.
  (I thought about including "this." in ufunc->uf_args, but I decided to go with this method because it seemed like it would have a wide range of effects.)
  (However, I think it would be good to challenge someone.)

**Comment by h-east** [comment]
Hmm, This patch seems to have a problem.
`fp->uf_arg_types[i]` is a pointer.  copy an address of `type`.
Therefore, changing the value of a member will affect others...

```
parse_argument_types(ufunc_T *fp, garray_T *argtypes, int varargs, garray_T *argflags)
{
    ....
    fp->uf_arg_types = ALLOC_CLEAR_MULT(type_T *, len);
    ....
    type = parse_type(&p, &fp->uf_type_list, TRUE);
    ....
    fp->uf_arg_types[i] = type;
    type = parse_type(&p, &fp->uf_type_list, TRUE);
    ....
    fp->uf_arg_types[i]->tt_flags &= ~TTFLAG_ARG_OBJM;              // BAD
    ....
}
```

**Comment by h-east** [comment]
Fix it!

change point
- I stopped checking in generate_CALL() and set the type of object member in fp->uf_arg_types[i] in parse_argument_types().

**Comment by yegappan** [comment]
> Fix it!
> 
> change point
> 
> * I stopped checking in `generate_CALL()` and set the type of object member to `fp->uf_arg_types[i]` 
> in `parse_argument_types()`.
>

It looks like the latest change impacts the compilation of arguments to every object method,
The type checking for object methods is working correctly.  I thought only the type checking
for arguments to the new() constructor function with the same name as a member variable
is not working.

**Comment by h-east** [comment]
@yegappan, @chrisbra 
Yes, I am going to start talking about the language specification of Vim9 script.
I think it is not good that object members can be used as formal arguments for methods other than class constructors.
Reading vim9class.txt, my guess is that Bram wanted to apply this mechanism only to class constructors.

Proposal:
Let's make it an error when an object member is specified as formal arguments to a class method that is not a constructor?

**Comment by h-east** [comment]
Currently, the following code is working. Is this good?

```vim
vim9script

class C
  this.str: string
  this.num: number
  def new(this.str, this.num)
  enddef
  def FuncA(this.str, this.num)
  enddef
endclass

def Check()
  try
    var c = C.new('a', 1)
	c.FuncA('xyz', 789)
	echo c.str
	echo c.num
  catch
    echo $'Exception was caught: {v:exception}'
  endtry
enddef

Check()
```

**Comment by yegappan** [comment]
> @yegappan, @chrisbra Yes, I am going to start talking about the language specification of Vim9 script. I think it is not good that object members can be used as formal arguments for methods other than class constructors. Reading vim9class.txt, my guess is that Bram wanted to apply this mechanism only to class constructors.
> 
> Proposal: Let's make it an error when an object member is specified as formal arguments to a class method that is not a constructor?

Yes.  Sounds good.  The compilation of an class or object method should fail under the following conditions:

1. The "this." prefix is used for an argument name,
2. The argument name overlaps with a class variable name.

Note that it is acceptable to use an argument name to an object method with the same
name as the object variable name without the "this." prefix.

**Comment by h-east** [comment]
First of all, could you please accept this patch?
After that, would you like to discuss the following suggestions on issues/#13152 ?

> Proposal:
Let's make it an error when an object member is specified as formal arguments to a class method that is not a constructor?

**Comment by yegappan** [comment]
> > But if the arguments to the FuncA()
> > method are changed to "(str, num)", without the "this." prefix, then the
> > compilation should succeed.
> 
> No. The following code will result in an error regardless of whether with my patch or not. Type specification is required if the formal argument is not an object member. This is correct. This behavior will remain the same in the future.
> 

Yes.  What I meant is that it is allowed to use a parameter name (with the type specification)
with the same name as an object variable name.

File: src/userfunc.c [hunk 1]
```diff
@@ -172,6 +174,7 @@ one_function_arg(
 		type = vim_strsave((char_u *)
 					    (is_vararg ? "list<any>" : "any"));
 	    ((char_u **)argtypes->ga_data)[argtypes->ga_len++] = type;
+	    ((int8_T *)is_argobjm->ga_data)[is_argobjm->ga_len++] = 0;
```
Comment (yegappan) [inline src/userfunc.c]: Can you rename "is_argobjm" to "is_arg_objm" or "arg_objm" (it is more clear)?

Comment (yegappan) [inline src/userfunc.c]: As "is_argobjm" is a list of Booleans, can you use "FALSE" instead of "0"?

File: src/userfunc.c [hunk 2]
```diff
@@ -350,11 +356,13 @@ get_function_args(
...
 		newargs->ga_len++;
 
-		if (argtypes != NULL && ga_grow(argtypes, 1) == OK)
+		if (argtypes != NULL && ga_grow(argtypes, 1) == OK
+			&& is_argobjm != NULL && ga_grow(is_argobjm, 1) == OK)
 		{
 		    // TODO: use the actual type
 		    ((char_u **)argtypes->ga_data)[argtypes->ga_len++] =
 						  vim_strsave((char_u *)"any");
+		    ((int8_T *)is_argobjm->ga_data)[is_argobjm->ga_len++] = 1;
```
Comment (yegappan) [inline src/userfunc.c]: As "is_argobjm" is a list of Booleans, can you use "TRUE" instead of "1"?

**Review by yegappan — CHANGES_REQUESTED** [review]

**Comment by h-east** [comment]
> Yes. What I meant is that it is allowed to use a parameter name (with the type specification)
with the same name as an object variable name.

Yes, I understand that.

This PR is a fix for the code below which unexpectedly caused an error.
(In addition, type checking is also performed when an object method (this.~) is specified as formal arguments in a method other than the constructor.)   <--- I would like to discuss this part later.

```vim
vim9script

class C
  this.str: string
  this.num: number
  def newFuncA(str: any, num: any)
  enddef
endclass

def Check()
  try
    var c = C.newFuncA(1, 2)
  catch
    echo $'Exception was caught: {v:exception}'
  endtry
enddef

Check()
```

File: src/userfunc.c (hunk 1 above)
Comment (h-east) [reply src/userfunc.c]: I'll rename to `arg_objm`.

**Review by yegappan — APPROVED** [review]

**Comment by chrisbra** [comment]
thanks!

**Comment by vim-ml** [comment]
Hi,

On Sun, Sep 24, 2023 at 6:44 AM Christian Brabandt <
***@***.***> wrote:

> thanks!
>
>
>
The CI ASAN test is failing after this patch.  There is a memory leak with
this PR.

Regards,
Yegappan

**Comment by h-east** [comment]
Hmm, memory leak not fixed.

**Pull request closed without merging**